segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,179 @@
+import java.util.function.Supplier;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Manages the Write Queue Concurrency.
+ */
+@ThreadSafe
+class ConcurrencyManager {
+    //region Members
+
+    @VisibleForTesting
+    static final double SIGNIFICANT_DIFFERENCE = 0.1;
+    @VisibleForTesting
+    static final long MIN_FREQUENCY_MILLIS = 1000;
+    @VisibleForTesting
+    static final long STALE_MILLIS = MIN_FREQUENCY_MILLIS * 4;
+    private final int minParallelism;
+    private final int maxParallelism;
+    private final Supplier<Long> timeSupplier;
+    @GuardedBy("this")
+    private long recentTotalWrittenLength;
+    @GuardedBy("this")
+    private int recentWriteCount;
+    private final AtomicReference<Snapshot> lastSnapshot;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the ConcurrencyManager class.
+     *
+     * @param minParallelism The minimum degree of parallelism desired.
+     * @param maxParallelism The maximum degree of parallelism desired.
+     */
+    ConcurrencyManager(int minParallelism, int maxParallelism) {
+        this(minParallelism, maxParallelism, System::nanoTime);
+    }
+
+    /**
+     * Creates a new instance of the ConcurrencyManager class.
+     *
+     * @param minParallelism The minimum degree of parallelism desired.
+     * @param maxParallelism The maximum degree of parallelism desired.
+     * @param timeSupplier   A Supplier that returns the current time, in nanoseconds.
+     */
+    @VisibleForTesting
+    ConcurrencyManager(int minParallelism, int maxParallelism, Supplier<Long> timeSupplier) {
+        this.timeSupplier = Preconditions.checkNotNull(timeSupplier, "timeSupplier");
+        Preconditions.checkArgument(minParallelism <= maxParallelism, "minParallelism must be <= maxParallelism.");
+        this.minParallelism = minParallelism;
+        this.maxParallelism = maxParallelism;
+        this.recentTotalWrittenLength = 0;
+        this.recentWriteCount = 0;
+        this.lastSnapshot = new AtomicReference<>();
+        resetSnapshot();
+    }
+
+    //endregion
+
+    //region Operations
+
+    /**
+     * Records the fact that a write with the given length has just completed successfully.
+     *
+     * @param writeLength The length of the write that completed.
+     */
+    void writeCompleted(int writeLength) {
+        synchronized (this) {
+            this.recentWriteCount++;
+            this.recentTotalWrittenLength += writeLength;
+        }
+    }
+
+    /**
+     * Gets a value indicating the current (most recent) degree of parallelism.
+     *
+     * @return The result
+     */
+    int getCurrentParallelism() {
+        return this.lastSnapshot.get().parallelism;
+    }
+
+    /**
+     * Updates the degree of parallelism (if needed), based on information collected since the last time this method was called.
```

Comment:
While working on this I experimented with a bunch of inputs (tput, latency, etc), and tried a few things. I noticed that, as long as throughput increases, latency increases very slightly (assuming constant write size), but the latency increase is insignificant compared to the throughput increase. When throughput plateaus, latency starts to increase at a much higher pace.. 

So I had the option of trying to optimize for two variables (throughput and latency), but that would make the whole algorithm a lot more complicated. Since throughput plateaus at a point, I could detect that much easier (using `SIGNIFICANT_DIFFERENCE`) and stop increasing concurrency.

Using the manual parallelism (what is currently there in `master`) as a benchmark, I observed that throughput alone can be used to achieve the same, if not better results.  Trying to fit latency in here did not add any extra value.